font/font.go

```diff
@@ -0,0 +1,304 @@
+func From(fnt Font, size Length) Font {
+	o := fnt
+	o.Size = size
+	return o
+}
+
+// Typeface identifies a particular typeface design.
+// The empty string denotes the default typeface.
+type Typeface string
+
+// Variant denotes a typeface variant, such as "Mono", "Smallcaps" or "Math".
+type Variant string
+
+// FontExtents contains font metric information.
+type FontExtents struct {
+	// Ascent is the distance that the text
+	// extends above the baseline.
+	Ascent Length
+
+	// Descent is the distance that the text
+	// extends below the baseline.  The descent
+	// is given as a positive value.
+	Descent Length
+
+	// Height is the distance from the lowest
+	// descending point to the highest ascending
+	// point.
+	Height Length
+}
+
+type FontFace struct {
+	Font Font
+	Face *opentype.Font
+}
+
+// Name returns a fully qualified name for the given font.
+func (f *FontFace) Name() string {
+	return f.Font.Name()
+}
+
+func (f *FontFace) FontFace(dpi float64) font.Face {
+	face, err := opentype.NewFace(f.Face, &opentype.FaceOptions{
+		Size: f.Font.Size.Points(),
+		DPI:  dpi,
+	})
+	if err != nil {
+		panic(err)
+	}
+	return face
+}
+
+// default hinting for OpenType fonts
+const defaultHinting = font.HintingNone
+
+// Extents returns the FontExtents for a font.
+func (f *FontFace) Extents() FontExtents {
+	var (
+		// TODO(sbinet): re-use a Font-level sfnt.Buffer instead?
+		buf  sfnt.Buffer
+		ppem = fixed.Int26_6(f.Face.UnitsPerEm())
+	)
+
+	met, err := f.Face.Metrics(&buf, ppem, defaultHinting)
+	if err != nil {
+		panic(fmt.Errorf("could not extract font extents: %v", err))
+	}
+	scale := f.Font.Size / Points(float64(ppem))
+	return FontExtents{
+		Ascent:  Points(float64(met.Ascent)) * scale,
+		Descent: Points(float64(met.Descent)) * scale,
+		Height:  Points(float64(met.Height)) * scale,
+	}
+}
+
+// Width returns width of a string when drawn using the font.
+func (f *FontFace) Width(s string) Length {
+	var (
+		pixelsPerEm = fixed.Int26_6(f.Face.UnitsPerEm())
+
+		// scale converts sfnt.Unit to float64
+		scale = f.Font.Size / Points(float64(pixelsPerEm))
+
+		width     = 0
+		hasPrev   = false
+		buf       sfnt.Buffer
+		prev, idx sfnt.GlyphIndex
+		hinting   = defaultHinting
+	)
+	for _, rune := range s {
+		var err error
+		idx, err = f.Face.GlyphIndex(&buf, rune)
+		if err != nil {
+			panic(fmt.Errorf("could not get glyph index: %v", err))
+		}
+		if hasPrev {
+			kern, err := f.Face.Kern(&buf, prev, idx, pixelsPerEm, hinting)
+			switch {
+			case err == nil:
+				width += int(kern)
+			case errors.Is(err, sfnt.ErrNotFound):
+				// no-op
+			default:
+				panic(fmt.Errorf("could not get kerning: %v", err))
+			}
+		}
+		adv, err := f.Face.GlyphAdvance(&buf, idx, pixelsPerEm, hinting)
+		if err != nil {
+			panic(fmt.Errorf("could not retrieve glyph's advance: %v", err))
+		}
+		width += int(adv)
+		prev, hasPrev = idx, true
+	}
+	return Points(float64(width)) * scale
```

Comment:
`return Points(width) * scale`

Reply:
`Points` is actually a `func` that takes a single `float64` argument.
so your proposed alternative wouldn't compile.